contracts/src/v0.7/dev/KeeperRegistryDev.sol

```diff
@@ -664,60 +625,149 @@ contract KeeperRegistryDev is
    */
   function getMaxPaymentForGas(uint256 gasLimit) public view returns (uint96 maxPayment) {
     (uint256 gasWei, uint256 linkEth) = getFeedData();
     uint256 adjustedGasWei = adjustGasPrice(gasWei, false);
     return calculatePaymentAmount(gasLimit, adjustedGasWei, linkEth);
   }
 
+  /**
+   * @notice retrieves the migration permission for a peer registry
+   */
+  function getPeerRegistryMigrationPermission(address peer) external view returns (MigrationPermission) {
+    return s_peerRegistryMigrationPermission[peer];
+  }
+
+  /**
+   * @notice sets the peer registry migration permission
+   */
+  function setPeerRegistryMigrationPermission(address peer, MigrationPermission permission) external onlyOwner {
+    s_peerRegistryMigrationPermission[peer] = permission;
+  }
+
+  function migrateUpkeeps(uint256[] calldata ids, address destination) external override {
+    require(
+      s_peerRegistryMigrationPermission[destination] == MigrationPermission.OUTGOING ||
+        s_peerRegistryMigrationPermission[destination] == MigrationPermission.BIDIRECTIONAL
+    );
+    require(address(s_transcoder) != address(0), "no transcoder");
+    require(ids.length > 0, "no IDs");
+    uint256 id;
+    Upkeep memory upkeep;
+    address admin;
+    uint256 totalBalanceRemaining;
+    address[] memory targets = new address[](ids.length);
+    uint32[] memory gasLimits = new uint32[](ids.length);
+    uint96[] memory balances = new uint96[](ids.length);
+    bytes[] memory checkDatas = new bytes[](ids.length);
+    bool isOwner = msg.sender == owner();
+    for (uint256 idx = 0; idx < ids.length; idx++) {
+      id = ids[idx];
+      upkeep = s_upkeep[id];
+      if (idx == 0) {
+        admin = upkeep.admin;
+      }
+      require(msg.sender == admin || isOwner);
+      require(upkeep.maxValidBlocknumber == UINT64_MAX, "upkeep must be active");
+      targets[idx] = upkeep.target;
+      gasLimits[idx] = upkeep.executeGas;
+      balances[idx] = upkeep.balance;
+      checkDatas[idx] = s_checkData[id];
+      totalBalanceRemaining = totalBalanceRemaining.add(upkeep.balance);
+      delete s_upkeep[id];
+      delete s_checkData[id];
+      s_upkeepIDs.remove(id);
+      emit UpkeepMigrated(id, upkeep.balance, destination);
+    }
+    s_expectedLinkBalance = s_expectedLinkBalance.sub(totalBalanceRemaining);
+    bytes memory encodedUpkeeps = abi.encode(admin, ids, targets, gasLimits, balances, checkDatas);
+    MigratableKeeperRegistryInterface(destination).receiveUpkeeps(
+      s_transcoder.transcodeUpkeeps(address(this), destination, encodedUpkeeps)
+    );
+    LINK.transfer(destination, totalBalanceRemaining);
+  }
+
+  function receiveUpkeeps(bytes calldata encodedUpkeeps) external override {
+    require(
+      s_peerRegistryMigrationPermission[msg.sender] == MigrationPermission.INCOMING ||
+        s_peerRegistryMigrationPermission[msg.sender] == MigrationPermission.BIDIRECTIONAL
+    );
+    (
+      address admin,
```

Comment:
admins can be different for auto migrate

Reply:
can you elaborate? @infiloop2

Reply:
When onwer migrates multiple upkeeps at once, they can have different admins?

Reply:
@RyanRHall I think suggestion is that admin should be an array, not a single admin for all received upkeeps.

Reply:
ahhh good catch @akhilchainani . that's definitely a bug.